tests/fail/tree-borrows/write-during-2phase.stderr

```diff
@@ -1,11 +1,28 @@
-error: Undefined Behavior: read access through <TAG> is forbidden because it is a child of <TAG> which is Disabled.
+error: Undefined Behavior: read access through <TAG> is forbidden because it is a child of <TAG> which does not have read nor write permissions.
   --> $DIR/write-during-2phase.rs:LL:CC
    |
 LL |     fn add(&mut self, n: u64) -> u64 {
-   |            ^^^^^^^^^ read access through <TAG> is forbidden because it is a child of <TAG> which is Disabled.
+   |            ^^^^^^^^^ read access through <TAG> is forbidden because it is a child of <TAG> which does not have read nor write permissions.
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the Tree Borrows rules it violated are still experimental
-   = note: BACKTRACE:
+help: the tag <TAG> that forbids this access was reborrowed here with read and write permissions
+  --> $DIR/write-during-2phase.rs:LL:CC
+   |
+LL |       let _res = f.add(unsafe {
+   |  ________________^
+LL | |         let n = f.0;
+LL | |         // This is the access at fault, but it's not immediately apparent because
+LL | |         // the reference that got invalidated is not under a Protector.
+LL | |         *inner = 42;
+LL | |         n
+LL | |     });
+   | |______^
+help: <TAG> lost read and write permissions due to a write access at offsets [0x0..0x8]
```

Comment:
This should tell a story, e.g.
- TAG was created here in Reserved state by a reborrow (name parent tag, maybe?)
- TAG later changed to Disabled state due to a Foreign write access at offsets...

I guess @saethlin will want us to avoid using special Tree Borrows terminology here, but I really don't see a good way to avoid that. Replacing state names by long permission descriptions doesn't really explain that much IMO, and this here currently describes Reserved as "with read and write permissions" which is confusing since that sounds like Active.

Maybe we have the `help` line be a short one using special terms, and then have a more detailed `note`/`help` below (or somewhere) that explains a bit more, and that people familiar with TB can skip?

Reply:
> Replacing state names by long permission descriptions doesn't really explain that much IMO

🤷 I think it helps. It also gets us out of discussing disabled children.

Really we have a Stroustrup's Rule problem here: Everyone finds the new terminology confusing _at first_, but then later they want a terse notation for it. I've noticed that the pickup rate for changes in Miri can be amazingly slow, some rather experienced unsafe code users didn't notice the new span-based diagnostics until 6 months after they released.

So however you want to navigate all that is good with me. If we're going to put up all the Tree Borrows terminology, there needs to be an accessible way for people to learn what those terms mean. https://perso.crans.org/vanille/treebor/ is a great breakdown of the whole model, but it is quite intimidating for a newcomer who is just trying to figure out why Miri doesn't like their new data structure idea.

I've long wanted a sort of glossary built from common patterns that explains why they are invalid and how to fix them. https://miri.saethlin.dev/tree-borrows/ub exists now and will be complete once the upload finishes. I will work on such a glossary eventually if someone else doesn't beat me to it.

Reply:
> Stroustrup's Rule

Definitely. But we also have the problem that there's a lot going on and trying to cram that all into a single sentence will not result in anything readable.

If you have good suggestions for how to avoid the state names, I am all ears. But the PR as-is describes Reserved as "having read and write permissions" and that's just misleading -- and the more correct "having read and future-write permissions" is no better than just saying "Reserved".

Reply:
> If you have good suggestions for how to avoid the state names, I am all ears.

I don't. I wish I did.

I was a bit surprised by the direction @Vanille-N took my comment, but as I dislike terse notation (because with my route through grad school I was constantly reading papers outside my subfield) I am not inclined to get in anyone's way when they are removing it. My initial comment was focused on the consistency or logical flow of terminology/phrasing because that's a clearer topic. If it turns out that addressing my comment that requires opening up a lot more lines of discussion let's do the incremental thing and add all the extra notes now and decide on the English that surrounds them later.

> We could also consider renaming Disabled to Inactive, Invalid, Deactivated, something like that.

Please not _another_ use of Invalid :p 

I don't particularly care what color the shed is, but I want us to be happy with the term before Tree Borrows becomes a default.